src/hotspot/share/logging/logAsyncWriter.cpp

```diff
@@ -42,28 +43,65 @@ class AsyncLogWriter::AsyncLogLocker : public StackObj {
   }
 };
 
-void AsyncLogWriter::enqueue_locked(const AsyncLogMessage& msg) {
-  if (_buffer.size() >= _buffer_max_size) {
+// LogDecorator::None applies to 'constant initialization' because of its constexpr constructor.
+const LogDecorations& AsyncLogWriter::None = LogDecorations(LogLevel::Warning, LogTagSetMapping<LogTag::__NO_TAG>::tagset(),
+                                      LogDecorators::None);
+
+// Reserve space for a flush token, so 'push_flush_token' always succeeds.
+AsyncLogWriter::Buffer::Buffer(size_t capacity) : _pos(0) {
+  _buf = NEW_C_HEAP_ARRAY(char, capacity, mtLogging);
+  _capacity = capacity - AsyncLogWriter::TOKEN_SIZE;
+}
+
+AsyncLogWriter::Buffer::~Buffer() {
+  FREE_C_HEAP_ARRAY(char, _buf);
+}
+
+bool AsyncLogWriter::Buffer::push_back(LogFileStreamOutput* output, const LogDecorations& decorations, const char* msg) {
+  size_t len = strlen(msg) + 1; // including trailing zero
+  size_t sz = align_up(sizeof(Message) + len, sizeof(void*));
+
+  if (_pos + sz <= _capacity || output == nullptr/*token*/) {
```

Comment:
Currently, there's only one token, ie. the flushing token. We use it to implement "flushing" semantic. 

`AsyncLogWriter::flush` inserts it and is blocked until "flushing" is complete. 
Therefore, one queue can only have one token anytime. I reserve it in ctor so enqueuing of the token always succeeds. Unlike regular enqueuing, failure is not acceptable because it will hang AsyncLogWriter thread. 

Previously, I have an object called 'Token'. Since now the constructor only works for in-place new, I can't construct it on on c-heap or stack, so I use a tuple {nullptr,  AsyncLogWriter::None, ""} here.

```
void AsyncLogWriter::Buffer::push_flush_token() {
  bool result = push_back(nullptr, AsyncLogWriter::None, "");
  assert(result, "fail to enqueue the flush token");
}
```

Reply:
Ah, I see. This is easy to miss. May I suggest a slight alteration that makes it easier to understand to the casual reader?

In Message, define a static function that calculates message byte size (the various definitions existing make me nervous, easy for them to drift apart):
```
    // Calculate size for a prospective Message object depending on its message length
    static constexpr size_t calc_size(int message_len) {
      return align_up(sizeof(Message) + message_len + 1, sizeof(void*));
    }
```

Use this in Message::size().

Get rid of TOKEN, and in the ctor, let _capacity be the real capacity. Instead, in push_back, do:

```
bool AsyncLogWriter::Buffer::push_back(LogFileStreamOutput* output, const LogDecorations& decorations, const char* msg) {
  const size_t sz = Message::calc_size(strlen(msg));
  const bool is_push_token = output == nullptr;
  // Always leave headroom for the push token. Pushing the push token must always succeed.
  const size_t headroom = (!is_push_token) ? Message::calc_size(0) : 0;
  if (_pos + sz <= (_capacity - headroom)) {
    new(_buf + _pos) Message(output, decorations, msg);
    _pos += sz;
    return true;
  }
  return false;
}
```

Now the magic is in exactly one place, and easy to understand with one glance. And we have one central place to calculate the message byte size.

Reply:
update it accordingly.